### PR TITLE
Auth/[pm-2759] - TDE - SSO and 2FA routing logic

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

- Add test suites for existing `SsoComponent` and `TwoFactorComponent` post SSO login redirect logic to ensure we can add to the logic without breaking existing flows 
- Refactor the two components to improve the code and ensure the two implementations of the post SSO redirect logic match each other (future work should probably extract this post SSO redirect logic into a service) 
- Add Trusted Device Encryption navigation logic into both components + tests for new scenarios 

## Code changes

- **file.ext:** TODO

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/angular/src/auth/components/sso.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  it("should create", () => {
+    expect(component).toBeTruthy();
+  });
+
+  describe("navigateViaCallbackOrRoute(...)", () => {
+    it("calls the provided callback when callback is defined", async () => {
+      const callback = jest.fn().mockResolvedValue(null);
+      const commands = ["some", "route"];
+
+      await (component as any).navigateViaCallbackOrRoute(callback, commands);
```
Comment (coroiu) [inline libs/angular/src/auth/components/sso.component.spec.ts:106]: suggestion 1: Remove tests for `navigateViaCallbackOrRoute`, you are already testing it below.

suggestion 2: Redefine `navigateViaCallbackOrRoute` as a standalone util function and test it separately. 

**Why**
Testing private functions usually indicates a code/design smell and is usually avoided. Tests should verify that the behavior of the outward interface meets expectations, and doesn't deviate from those expectations with future changes. In those cases private functions are not part of the outward interface, and are therefore usually seen as "implementation detail"

File: libs/angular/src/auth/components/sso.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+  successRoute: string;
+  trustedDeviceEncRoute: string;
+  changePasswordRoute: string;
+  forcePasswordResetRoute: string;
+  logIn(code: string, codeVerifier: string, orgIdFromState: string): Promise<AuthResult>;
+  handleLoginError(e: any): Promise<void>;
+}
+
+// The ideal scenario would be to not have to test the protected / private methods of the SsoComponent
+// but that will require a refactor of the SsoComponent class is out of scope for now.
```
Comment (coroiu) [inline libs/angular/src/auth/components/sso.component.spec.ts:39]: nit/typo:
```suggestion
// but that will require a refactor of the SsoComponent class which is out of scope for now.
```

File: libs/angular/src/auth/components/sso.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+  const code = "code";
+  const codeVerifier = "codeVerifier";
+  const orgIdFromState = "orgIdFromState";
+
+  // Mock component callbacks
+  let mockOnSuccessfulLogin = jest.fn();
+  let mockOnSuccessfulLoginNavigate = jest.fn();
+  let mockOnSuccessfulLoginTwoFactorNavigate = jest.fn();
+  let mockOnSuccessfulLoginChangePasswordNavigate = jest.fn();
+  let mockOnSuccessfulLoginForceResetNavigate = jest.fn();
```
Comment (coroiu) [inline libs/angular/src/auth/components/sso.component.spec.ts:71]: **suggestion:** test setups like these should ideally be performed inside `beforeEach`, i.e. the variables are defined here, but assigned in `beforeEach`. This is to avoid cross-test behavior "contamination". It's an edge case and increases the files size I know :(

**Background**
in an older version of the Unit test conventions document I actually encouraged putting some variables outside of `beforeEach` 

image

But I got comments on it and eventually removed it

image

File: libs/angular/src/auth/components/sso.component.ts [hunk 4]
```diff
@@ -179,90 +179,157 @@ export class SsoComponent {
...
-      } else if (response.forcePasswordReset !== ForceResetPasswordReason.None) {
-        if (this.onSuccessfulLoginForceResetNavigate != null) {
-          await this.onSuccessfulLoginForceResetNavigate();
-        } else {
-          this.router.navigate([this.forcePasswordResetRoute]);
-        }
+      // User must set password if they don't have one and they aren't using either TDE or key connector.
+      const requireSetPassword =
+        !acctDecryptionOpts.hasMasterPassword &&
+        acctDecryptionOpts.keyConnectorOption === undefined;
```
Comment (coroiu) [inline libs/angular/src/auth/components/sso.component.ts:200]: question: your comment here suggests the following logic

```!hasMasterPassword && !(usingKeyConnector || usingTde)```

But the code only checks

```!hasMasterPassword && !usingKeyConnector```

Am I missing something?

File: libs/angular/src/auth/components/sso.component.ts [hunk 5]
```diff
@@ -179,90 +179,157 @@ export class SsoComponent {
...
+    } else {
+      // Navigate to TDE page (if user was on trusted device and TDE has decrypted
+      //  their user key, the lock guard will redirect them to the vault)
+      this.router.navigate([this.trustedDeviceEncRoute], {
+        queryParams: {
+          identifier: orgIdentifier,
+        },
+      });
+    }
+  }
```
Comment (coroiu) [inline libs/angular/src/auth/components/sso.component.ts:274]: suggestion(non-blocking): write using guard statements

```suggestion
  private async handleTrustedDeviceEncryptionEnabled(
    authResult: AuthResult,
    orgIdentifier: string,
    acctDecryptionOpts: AccountDecryptionOptions
  ): Promise<void> {
    // If user doesn't have a MP, but has reset password permission, they must set a MP
    if (
      !acctDecryptionOpts.hasMasterPassword &&
      acctDecryptionOpts.trustedDeviceOption.hasManageResetPasswordPermission
    ) {
      // Change implies going no password -> password in this case
      return await this.handleChangePasswordRequired(orgIdentifier);
    } 
    
    if (authResult.forcePasswordReset !== ForceResetPasswordReason.None) {
      return await this.handleForcePasswordReset(orgIdentifier);
    } 
    
    // Navigate to TDE page (if user was on trusted device and TDE has decrypted
    //  their user key, the lock guard will redirect them to the vault)
    this.router.navigate([this.trustedDeviceEncRoute], {
      queryParams: {
        identifier: orgIdentifier,
      },
    });
  }
```

File: libs/angular/src/auth/components/two-factor.component.ts [hunk 6]
```diff
@@ -201,55 +207,125 @@ export class TwoFactorComponent extends CaptchaProtectedComponent implements OnI
...
+  }
+
+  private async handleLoginResponse(authResult: AuthResult) {
+    if (this.handleCaptchaRequired(authResult)) {
       return;
     }
+
+    // Note: calling this here is different from the SSO component.
+    // The SsoComponent only executes this logic as part of the handleSuccessfulLogin logic.
+    await this.handleOnSuccessfulLogin();
```
Comment (coroiu) [inline libs/angular/src/auth/components/two-factor.component.ts:222]: question: can you clarify this a bit? What is different, why, and what impact does that have (why does it matter)?

File: libs/angular/src/auth/components/two-factor.component.ts [hunk 7]
```diff
@@ -201,55 +207,125 @@ export class TwoFactorComponent extends CaptchaProtectedComponent implements OnI
...
-          queryParams: {
-            identifier: this.identifier,
-          },
-        });
-      }
+  private async handleSuccessfulLogin() {
+    await this.navigateViaCallbackOrRoute(this.onSuccessfulLoginNavigate, [this.successRoute]);
+  }
+
+  private async navigateViaCallbackOrRoute(
```
Comment (coroiu) [inline libs/angular/src/auth/components/two-factor.component.ts:320]: **thought:** It seem that `TwoFactorComponent` and `SsoComponent` are very similar and that we are duplicating logic between them. Not something that needs to be fixed here though, just an observation. Did you create a ticket for further refactors of these components?

File: libs/angular/src/auth/components/sso.component.ts [hunk 8]
```diff
@@ -179,90 +179,157 @@ export class SsoComponent {
...
+          authResult,
+          orgIdentifier,
+          acctDecryptionOpts
+        );
+      } else if (requireSetPassword) {
+        // Change implies going no password -> password in this case
+        await this.handleChangePasswordRequired(orgIdentifier);
+      } else if (authResult.forcePasswordReset !== ForceResetPasswordReason.None) {
+        await this.handleForcePasswordReset(orgIdentifier);
       } else {
```
Comment (coroiu) [inline libs/angular/src/auth/components/sso.component.ts:219]: **suggestion:** Could you maybe explain the difference between these two somehow?

File: libs/angular/src/auth/components/sso.component.ts [hunk 9]
```diff
@@ -179,90 +179,157 @@ export class SsoComponent {
     return authorizeUrl;
   }
 
-  private async logIn(code: string, codeVerifier: string, orgIdFromState: string) {
+  private async logIn(code: string, codeVerifier: string, orgIdentifier: string) {
```
Comment (coroiu) [inline libs/angular/src/auth/components/sso.component.ts:182]: **suggestion(non-blocking):** I like the refactor! Now that it's a bit more simplified, could we maybe convert the if-statements to be written as guard-statements instead?

**Review by coroiu — COMMENTED** [review]
Man these components have complex logic. There is so much domain knowledge shoved into these that it's hard to grasp changes to the function as a whole. However, given that this passes through QA and the fact that we now have unit tests should give us a lot more confidence to make changes and improve this in the future! Great work!

Also, nice
image

File: libs/angular/src/auth/components/two-factor.component.ts (hunk 7 above)
Comment (trmartin4) [reply libs/angular/src/auth/components/two-factor.component.ts:320]: That’s on my to-do list.

File: libs/angular/src/auth/components/sso.component.spec.ts (hunk 3 above)
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/sso.component.spec.ts:71]: Thank you for the great callout!

File: libs/angular/src/auth/components/sso.component.ts (hunk 4 above)
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/sso.component.ts:200]: Sorry about the lack of clarity. I've hopefully resolved that. 
```
// In the standard, non TDE case, a user must set password if they don't
// have one and they aren't using key connector.
// Note: TDE & Key connector are mutually exclusive org config options.
```

File: libs/angular/src/auth/components/two-factor.component.ts (hunk 7 above)
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/two-factor.component.ts:320]: Yes, we hopefully can refactor these similar post SSO login redirect pathways into leveraging a common service in the future to improve code re-use. Just the timeframe involved is out of scope for now. 

Thank you, @trmartin4, for capturing that tech debt task.

File: libs/angular/src/auth/components/sso.component.ts (hunk 8 above)
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/sso.component.ts]: image
Hopefully that made it clearer!

**Review by coroiu — APPROVED** [review]
💯

File: libs/angular/src/auth/components/two-factor.component.ts [hunk 10]
```diff
@@ -201,55 +207,137 @@ export class TwoFactorComponent extends CaptchaProtectedComponent implements OnI
...
+    });
+  }
+
+  private async handleSuccessfulLogin() {
+    await this.navigateViaCallbackOrRoute(this.onSuccessfulLoginNavigate, [this.successRoute]);
+  }
+
+  private async navigateViaCallbackOrRoute(
+    callback: () => Promise<unknown>,
+    commands: any[],
```
Comment (justindbaur) [inline libs/angular/src/auth/components/two-factor.component.ts:334]: `unknown[]` if possible.

**Review by justindbaur — COMMENTED** [review]
Just one tiny change if possible.

File: libs/angular/src/auth/components/two-factor.component.ts (hunk 10 above)
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/two-factor.component.ts:334]: Are we sure we want to use unknown here when Angular's navigate accepts an `any[]`?
image

Comment (justindbaur) [reply libs/angular/src/auth/components/two-factor.component.ts:334]: Yep, `unknown` will still naturally cast to `any` there but would deter us from trying to use the `commands` in our own method in the future because you'd have to cast up to something else to do so.

**Review by justindbaur — APPROVED** [review]
Thanks!

**Review by jlf0dev — APPROVED** [review]

File: libs/angular/src/auth/components/two-factor.component.ts (hunk 7 above)
Comment (trmartin4) [reply libs/angular/src/auth/components/two-factor.component.ts:332]: I've captured that in https://bitwarden.atlassian.net/browse/PM-3095, which I put at the top of our tech debt backlog.
